[Docs] Add intro and quickstart docs (#859)
mui/mui-toolpad: docs/data/toolpad/getting-started/quickstart.md

# Quickstart

> <p class="description">Get started building a Toolpad app.</p>

Reviewer: ```suggestion
<p class="description">Learn the fundamentals of building with Toolpad by creating and deploying a simple app.</p>
```

> This page will guide you into creating your very first Toolpad application. You will create a "hello world" type of application that will ask for a name and print a greeting.

Reviewer: ```suggestion
This guide will walk you through the process of creating a basic Toolpad application.
You'll use the MUI X DataGrid component to display a list of dog breeds from the [Dog API](https://dog.ceo/dog-api/).
When you click on the name of a breed, its photo will be displayed using the Material UI Image component.
```

> 

Reviewer: I would suggest adding this section to give the reader more info about what they'll learn here.

```suggestion

## Purpose

This Quickstart guide is intended to introduce you to the fundamentals of building with Toolpad.
By the end, you should be able to:

- set up a new Toolpad app
- navigate through your workspace
- add components, data sources, and queries to an app
- bind data sources and components
- deploy a Toolpad app
```

> ## Prerequisites

Reviewer 2: > can we assume that Toolpad users won't be intimidated by some JavaScript?

I think so yes, atm you won't get far without it

> Make sure you have a running Toolpad instance.

Reviewer: When I run the commands as shown here, I get this error:

`ERROR: .FileNotFoundError: [Errno 2] No such file or directory: './docker/compose/docker-compose.yml'`

What am I doing wrong? Is there something I'm missing here? (I'm clueless when it comes to Docker but I do have it installed and running.)

Reviewer 3: https://github.com/mui/mui-toolpad/pull/859#discussion_r968851183

1. Download the docker compose file

   ```sh
   curl -LO https://raw.githubusercontent.com/mui/mui-toolpad/master/docker/compose/docker-compose.yml
   ```

1. Start the docker compose services

   ```sh

>    docker-compose -f docker/compose/docker-compose.yml up -d

Reviewer 3: Based after https://github.com/mui/mui-toolpad/pull/859/files#r968773969, I could reproduce.

```suggestion
   docker-compose -f docker-compose.yml up -d
```

Reviewer 3: This is **off-topic** for this PR but it could be interesting for the onboarding experience:

1. I wonder if it would make sense to ask to create an empty folder.
2. The amd64 emulator is super slow on a MacBook M1/M2

<img width="315" alt="Screenshot 2022-09-12 at 21 48 10" src="https://user-images.githubusercontent.com/3165635/189742558-e1cabd23-0e5d-44c8-ba36-7e8a36d1f348.png">

It's not usable, e.g.

<img width="743" alt="Screenshot 2022-09-12 at 21 49 39" src="https://user-images.githubusercontent.com/3165635/189742830-e2349790-d05a-4134-90c7-040f29b990cd.png">

3. I don't have the drop component here

<img width="897" alt="Screenshot 2022-09-12 at 21 49 58" src="https://user-images.githubusercontent.com/3165635/189742892-4672a262-29f1-4f26-ab19-589f2b63888d.png">

Author: Created this issue so we can keep track of it

Reviewer 2: We could pick this back up at some point: https://github.com/mui/mui-toolpad/pull/618

I'm wondering if it's the real problem though. I'm on M1 and don't see the same behavior. appsmith and Toolpad hum along on Docker just fine for me.

<p class="description">
This page will guide you into creating your very first Toolpad application. We will create a "hello world" type of application that will ask for a name and print a greeting.
</p>

## Prerequisites

Make sure you have a running Toolpad instance.

1. Download the docker compose file

   ```sh
   curl -LO https://raw.githubusercontent.com/mui/mui-toolpad/master/docker/compose/docker-compose.yml
   ```

1. Start the docker compose services

   ```sh
   docker-compose -f docker/compose/docker-compose.yml up -d
   ```

Toolpad will be accessible under `http://localhost:3000/`.

## Building your first application

1. Open Toolpad by navigating to [http://localhost:3000/](http://localhost:3000/)

   ![Apps overview](/static/toolpad/apps-overview.png)

1. Click "CREATE NEW" and name your application. Confirm by clicking "CREATE"

   ![Toolpad overview](/static/toolpad/step-1.png)

1. Hover over **Component library** and drag **DataGrid** and **Image** components into **Canvas**

   ![Drag components](/static/toolpad/step-2.png)

   Congratulations, we are done building the UI! Now all we need to do is connect it with a data source.

> 1. Click anywhere inside **Canvas** (except on the components that you just added) to deselect added components

Reviewer 2: This is a little UX problem that is probably worth taking a look at. To get to page level options is not very discoverable. I feel like having to deselect something before you can get to the UI to edit something else isn't optimal.
@prakhargupta1 @gerdadesign I believe you already started an effort on this with the bottom panel that you started to design?

# Quickstart

> <p class="description">Learn the fundamentals of building with Toolpad by creating and deploying a simple app.</p>

Reviewer 2: In terms of nomenclature I see "app" being used as well as "application". Shall we standardize on "application", (or "Toolpad application" when we need to be more specific)

```suggestion
<p class="description">Learn the fundamentals of building with Toolpad by creating and deploying a simple application.</p>
```

# Quickstart

Get started building a Toolpad app!

<p class="description">
This page will guide you into creating your very first Toolpad application. We will create a "hello world" type of application that will ask for a name and print a greeting.
</p>

> ## Prerequisites

Reviewer 2: Good to have this section here, we should probably add a "setup" page or something with more elaborate installation instructions

Author: I agree, as I mentioned in some other comment I think in the future we should add something like "Provisioning" section which explains in detail how to setup your own self-hosted version and not just run demo example

Make sure you have a running Toolpad instance.

1. Download the docker compose file

   ```sh
   curl -LO https://raw.githubusercontent.com/mui/mui-toolpad/master/docker/compose/docker-compose.yml
   ```

1. Start the docker compose services

   ```sh
   docker-compose -f docker/compose/docker-compose.yml up -d
   ```

Toolpad will be accessible under `http://localhost:3000/`.

> ## Navigating Toolpad

Reviewer 2: Does it make sense to make this part of the quickstart tutorial? 
If I understand correctly, the reason it's put here is so that we can refer to the sections when explaining how to do an action. Maybe it's better to just make the screenshots clearer? Perhaps with arrows? In a later stage maybe even animating them?

Author: > If I understand correctly, the reason it's put here is so that we can refer to the sections when explaining how to do an action

That's exactly the reason why I put it there

> Maybe it's better to just make the screenshots clearer? Perhaps with arrows?

Do you mean like don't use overlays with text but put something like rectangular shapes and point to the them with arrow and write a name of a section outside of the screenshot?

Also I was thinking that maybe @gerdadesign could take over this and make it "right", I just wanted to do a first pass to have something to use 😅

Reviewer 2: > That's exactly the reason why I put it there

I think there could be value in making our quickstart easy enough to follow without needing to run the reader through a glossary of terms. We can do that by, instead of describing it, making the screenshots in the steps indicate what is to be clicked.

Author: I'd prefer to "teach" users terminology and then use text to refer to the actual section. While screenshots might be more explicit, they are also harder to maintain, so if that makes it possible to minimise amount of screenshots - I'd suggest to go that way

Reviewer 2: Yep, I think that makes sense for general documentation. But we're talking about a quickstart tutorial here. I feel like the primary goal of this section would be to get users as quickly as possible to that dopamine hit of having built a working application. It will give them an overall feel of what it's like building an application, as well as a starting point to go tinkering with it on their own. I'm wondering whether too much theory in here could put off users finishing the tutorial.

Author: You are right 👍 I moved it to `Overview` (previously introduction) section

![Toolpad app](/static/toolpad/app-ui-chrome.png)

There are 4 most frequently used areas:

- `Instance editor` - this is there you define and manage pages, connection configurations, custom component definitions, etc...

> - `Component's library` - drag component instance from this drawer and drop it in `Application editor` to build UI

Reviewer 4: ```suggestion
- `Components library` - drag component instance from this drawer and drop it in `Application editor` to build UI
```

Author: I'll update these names once we conclude UI Terminology RFC and I will update it all across documentation

- `Application editor` - this is a place where you build your app and immediately preview how it looks like
- `Properties editor` - modify properties of a different instance

## Building your first application

1. Open Toolpad by navigating to [http://localhost:3000/](http://localhost:3000/)

>    ![Apps overview](/static/toolpad/apps-overview.png)

Reviewer 2: This screenshot seems broken. 

Aside, does it make sense to add one more level in the `/public/static/toolpad` folder? I feel like it could quickly grow out of proportions.

1. Click "CREATE NEW" and name your application. Confirm by clicking "CREATE"

   ![Toolpad overview](/static/toolpad/step-1.png)

1. Hover over `Component's library` and drag `DataGrid` and `Image` components into `Application editor`

   ![Drag components](/static/toolpad/step-2.png)

>    Congratulations, we are done building UI! Now all we need is to connect it with same data source

Reviewer 2: ```suggestion
   Congratulations, we are done building the UI! Now all we need to do is connect it with a data source.
```

1. Click anywhere inside `Application editor` (except on the components that you just added) to deselect added components

1. Locate `ADD QUERY` button inside `Properties editor` and press that to start configuring our data source

   ![Add query](/static/toolpad/step-3.png)

1. This time we are going to use simple `FETCH` type to query our data. Choose and press `CREATE QUERY`

   ![Fetch type](/static/toolpad/step-4.png)

> 1. Let's fetch some data about dogs from `https://dog.ceo/dog-api` (if you're more of a cat person you could use cat data API from `https://catfact.ninja`. We apologize in advance - following tutorial steps are more suited for the dog data API but should be easily adaptable for cats as well!)

Reviewer 2: Do we want to rely on these sort of APIs that we don't control? Perhaps we should just upload a static JSON with dummy data to github and fetch it through `https://raw.githubusercontent.com/`?

Author: I've removed sentence with cat's api ref. I'd like to keep dogs example for now as it requires to update multiple screenshots as well. I will add a follow up task to change this to something that we can control easier 👍

   ![Connected data](/static/toolpad/step-9.png)

1. Let's make our app a bit more interactive. We want to display an image of a selected breed. Let's create a dynamic query which reacts to the selection inside `DataGrid` component

   - `ADD QUERY` -> create `FETCH` type
   - `name` -> `imageQuery`
   - Add new parameter named `breed`

   ![Breed parameter](/static/toolpad/step-10.png)

   - Bind `breed` parameter value to `dataGrid.selection ? dataGrid.selection[0] : 'akita'` (grabs selected value from `dataGrid` or defaults to `akita` breed)

   ![Breed binding](/static/toolpad/step-11.png)

   - Then bind query `url` property to `https://dog.ceo/api/breed/${query.breed}/images/random`

   ![url binding](/static/toolpad/step-12.png)

   - Last `transform responese` using `return data.message;` and `SAVE`

1. Next we want to display a picture of a selected breed using `Image` component

   - select `Image` component

   - bind `src` prop to `imageQuery.data` and voila - we can preview pictures of a selected breed

   ![Preview image](/static/toolpad/step-13.png)

   Congratulations! 🎉 we now have an app fetching data from remote source and reacting to the user input!

> ## Deploying your application

Reviewer 2: This flow has changed a lot since https://github.com/mui/mui-toolpad/pull/875
